### ISSUE TITLE
Meet a bug seems type inference error

### ISSUE THREAD

**Issue by DancingCorpse** [ExtendedXmlSerializer/home#352]
Hi, I meet an "InvalidCastException" when deserializing the serialized content to my data model, it seems that there is incorrect type inference, this is the details of the exception:

> System.InvalidCastException
  HResult=0x80004002
  Message=Unable to cast object of type 'ExtendedXmlSerializerTest.Backend' to type 'ExtendedXmlSerializerTest.WorkerBackend'.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

and this is my data model:
```namespace ExtendedXmlSerializerTest
{
	public class Backend
	{
		public string Host { get; set; } = "localhost";
		public int Port { get; set; } = 8080;
	}
	public class Portal
	{
		public string Name { get; set; } = "portal";
		public Backend Backend { get; set; } = new Backend();
	}
	public class WorkerBackend
	{
		public string ServerUrl { get; set; } = "http://localhost:8080";
	}
	public class Worker
	{
		public string Name { get; set; } = "worker";
		public WorkerBackend Backend { get; set; } = new WorkerBackend();
	}

	public class Config
	{
		public Portal Portal { get; set; } = new Portal();
		public Worker Worker { get; set; } = new Worker();
	}
}
```

my serializer implementation:
```using ExtendedXmlSerializer.Configuration;
using ExtendedXmlSerializer.ExtensionModel.Content;
using ExtendedXmlSerializer.ExtensionModel.Xml;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Xml;

namespace ExtendedXmlSerializerTest
{
	public static class ConfigSerializer
	{
		public static Config LoadConfig(string filePath)
		{
			IExtendedXmlSerializer serializer = new ConfigurationContainer()
				.Ignore(typeof(List<>).GetProperty(nameof(List<object>.Capacity)))
				.UseOptimizedNamespaces()
				.Create();
			using (var stream = new StreamReader(filePath))
			{
				return serializer.Deserialize<Config>(stream);
			}
		}
		public static void SaveConfig(Config config, string filePath)
		{
			var serializer = new ConfigurationContainer()
				.Ignore(typeof(List<>).GetProperty(nameof(List<object>.Capacity)))
				.UseOptimizedNamespaces()
				.Create();
			var content = serializer.Serialize(new XmlWriterSettings
			{
				NewLineChars = Environment.NewLine,
				Indent = true,
				IndentChars = "  ",
				Encoding = Encoding.UTF8,
				DoNotEscapeUriAttributes = true
			}, config);
			File.WriteAllText(filePath, content);
		}
	}
}
```
and this is my test code:
```
static class Program
	{
		static void Main(string[] args)
		{
			Console.Write("input file name: ");
			var fileName = Console.ReadLine();
			var config = new Config();
			ConfigSerializer.SaveConfig(config, fileName);
			var config2 = ConfigSerializer.LoadConfig(fileName);
			Console.WriteLine($"portal backend: http://{config2.Portal.Backend.Host}:{config2.Portal.Backend.Port}");
			Console.WriteLine($"worker backend: {config2.Worker.Backend}");
		}
	}
```

Could anybody help me whether I'm using it the wrong way or something I missed?
Thanks.

**Comment by DancingCorpse** [comment]
The version I used is:

>     <PackageReference Include="ExtendedXmlSerializer" Version="2.1.20" />

**Comment by Mike-EEE** [comment]
Hi @DancingCorpse (LOL nice name 😁) thanks for writing in.

I did take a look at your scenario, and everything seems to be working as expected as demonstrated by this passing test here:

https://github.com/ExtendedXmlSerializer/home/blob/2a7f7effd9f02d797918655026d6547582524866/test/ExtendedXmlSerializer.Tests.ReportedIssues/Issue352Tests.cs#L12-L50

Have you had the opportunity to try out version 3.0 yet?

https://www.nuget.org/packages/ExtendedXmlSerializer/

```
Install-Package ExtendedXmlSerializer -Version 3.0.2
```

Perhaps that might treat you better. :)  Please let me know if you continue to run into any issues and I will take a look into it for you. 👍

**Comment by DancingCorpse** [comment]
Hi @Mike-EEE, thanks for the quick response.
The issue still exists when using the version 3.0.2. 
I modified the definition of the data model "Worker" to
```
public class Worker
{
	public string Name { get; set; } = "worker";
	public WorkerBackend WBackend { get; set; } = new WorkerBackend();
}
```
and it works fine, so I think the serializer maybe mixed up the model "Backend" and "WorkerBackend", one of them is used by "Portal" and the other is used by "Worker".

**Comment by Mike-EEE** [comment]
Hm weird @DancingCorpse ... are you able to provide what the generated XML looks like so I can take a look into this further for you?

**Comment by Mike-EEE** [comment]
Also, if you can create a repository that reproduces this error I can grab it and take a look into it.

The tricky part about using `File.Save` and `File.Load` is that it adds a lot of complexity to testing so we do not use it if we can help it.

**Comment by DancingCorpse** [comment]
Hi @Mike-EEE, the generated XML file content is
``` XML
<?xml version="1.0" encoding="utf-8"?>
<Config xmlns="clr-namespace:ExtendedXmlSerializerTest;assembly=ExtendedXmlSerializerTest">
  <Portal>
    <Name>portal</Name>
    <Backend>
      <Host>localhost</Host>
      <Port>8080</Port>
    </Backend>
  </Portal>
  <Worker>
    <Name>worker</Name>
    <Backend>
      <ServerUrl>http://localhost:8080</ServerUrl>
    </Backend>
  </Worker>
</Config>
```

I ran your unit test case and it did works, I'm not sure what's the difference.

**Comment by Mike-EEE** [comment]
Yeah let's see if you can get a `csproj` posted to a repository here on GitHub with the above code failing and I will look into it further for you @DancingCorpse. 👍

**Comment by DancingCorpse** [comment]
Hi @Mike-EEE, I created a test repo "https://github.com/DancingCorpse/ExtendedXmlSerializerTest.git" on which you can try it, thanks very much.

**Comment by Mike-EEE** [comment]
Yes indeed... looks like there's something going on with the type resolution.  In our tests we usually put the model classes under test as nested classes, which is probably the differentiating factor here.  Taking a look into this now.

**Comment by Mike-EEE** [comment]
Boo... this is not pretty.  Our reader situation is not the greatest.  We have a special attribute in place for members to delineate between type members and for when a child element is actually a list element, as classes can be a collection with properties as well.

Anyways, I had to extend this dubious practice to fix this issue.  Essentially, we were getting a conflict for members with the same name as a type in the class.  We should have picked this up much sooner but because we use nested classes (and they get a new name altogether) this does not satisfy this condition.

TBH these sorts of conditions well irritate me.  I am disappointed we have to employ such measures to fix these types of issues.  If I saw this in another project I would simply sigh.  So I guess don't throw rocks in glass houses or however that expression goes.

The reader should be much more robust.  There should be reading done in a property context, a collection context, and a collection property context.  None of that exists, and would take extensive effort to redo all of this, so this is the best we can do for now while preserving all legacy functionality.

Long story short @DancingCorpse you can check out the latest version here to see if it fixes your issue:

https://github.com/ExtendedXmlSerializer/home/pull/354#issuecomment-574123928

If it does we will push out a new version to NuGet later today.

**Comment by DancingCorpse** [comment]
Hi @Mike-EEE, the package of version 3.0.2.8-qcjowwvf works well, one problem still exists is that my original serialized XML content still can not be deserialized successfully, but it's not a big problem and I can handle it I think, really appreciate for your help.

**Comment by Mike-EEE** [comment]
Ahhh you mean now because of the new `exs:member=""` tag?  Yeah I am not sure what to say here, it's ugly.  Our XmlReader needs a rewrite and my time is limited these days. :(

One thing you can do is to consider using XSLT to put in that tag before sending it to ExtendedXmlSerializer.

**Comment by Mike-EEE** [comment]
Alright I totally hate this "fix" but again it will have to do for now.  Someday when I have a billion years to devote to rewriting this project I can gut out this reader and make it good.  Or someone can submit a PR and do it for me. 😆

Anyways, this is now available on NuGet:
https://www.nuget.org/packages/ExtendedXmlSerializer/

**Comment by DancingCorpse** [comment]
@Mike-EEE  Really appreciate it.
